### PR TITLE
Nest all related test fixtures under a single outer fixture

### PR DESCRIPTION
**Pull request by ssoloff**
This PR is in response to @DanVanAtta's [comment](https://github.com/triplea-game/triplea/pull/1718#pullrequestreview-41030791) that there should be a 1:1 correspondence between a Java source file and its corresponding test source file.

I had previously created three separate fixtures for testing features of the `DownloadUtils` class, and those fixtures resided in their own source files.  The fixtures were sufficiently complex that combining them all into a single fixture just to achieve a single test source file would have resulted in code that was difficult to read due to so many concerns being present in a single test class.  (I had also previously used the same pattern for the `MapDownloadController` tests, which are addressed in this PR, as well.)

The solution employed in this PR is to keep the fixture for each feature independent but to nest them within a single outer fixture that represents the container for all tests for a single class.  This allows us to achieve the goal of having a single test file per top-level Java class while allowing for multiple independent fixtures per class.  The implementation uses the JUnit `Enclosed` test runner on the outer fixture while leaving the test runners for the inner fixtures up to the test author.

Unfortunately, using the `Enclosed` test runner causes each test to be executed twice under certain conditions.  This is due to the `Enclosed` test runner being essentially a JUnit `Suite`.  The test discovery protocol of certain tools will identify the suite and the test classes that make up the suite independently.  Thus, those tests nested within an `Enclosed` test runner will be executed twice if not compensated for.

Please see the commit message for the details of how this may affect running tests from the Eclipse IDE (TL;DR it doesn't if you normally right-click a test file and Run As > JUnit Test).

To fix the problem when running tests from Gradle, I had to specifically exclude `Enclosed` suites from the Gradle test discovery protocol.  To do this, a unique naming convention for identifying such tests had to be established.  The convention chosen in this change is to use the suffix `Tests` instead of the more traditional `Test` for `Enclosed` suites.  Gradle is then directed to exclude tests whose class names end in `Tests`.  Thus, as long as new `Enclosed` suites follow this naming convention, no future changes need be made to the Gradle configuration to prevent duplicate test runs.

A screenshot of the Gradle test report for the package containing the tests changed in this PR is below.  This should give you an idea of how the results are presented.

nested-test-results

#### Issues for review

* Does the fact that a naming convention is required for `Enclosed` tests disqualify this approach?  (If the naming convention isn't followed, the worst case scenario is the nested tests will be run twice.)
* If a naming convention is acceptable, is `Tests` an appropriate suffix?  Other alternatives I considered were `Suite` and `TestSuite`, but `Tests` is what is used in the `Enclosed` example code.

### CODE REVIEW

File: src/test/java/games/strategy/engine/framework/map/download/DownloadUtilsTests.java [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+import org.junit.experimental.runners.Enclosed;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import games.strategy.engine.framework.map.download.DownloadUtils.DownloadLengthSupplier;
+
+@RunWith(Enclosed.class)
+public final class DownloadUtilsTests {
```
Comment (ssoloff) [inline src/test/java/games/strategy/engine/framework/map/download/DownloadUtilsTests.java]: An example of an `Enclosed` suite.  This is effectively a fixture that groups all related fixtures for a single class.  In this case, `DownloadUtils`.

File: src/test/java/games/strategy/engine/framework/map/download/DownloadUtilsTests.java [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+  private static final String URI = "some://uri";
```
Comment (ssoloff) [inline src/test/java/games/strategy/engine/framework/map/download/DownloadUtilsTests.java]: The outer fixture can contain `static` fields and methods that may be useful to the nested fixtures.

File: src/test/java/games/strategy/engine/framework/map/download/DownloadUtilsTests.java [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+  @RunWith(MockitoJUnitRunner.class)
+  public static final class DownloadToFile {
```
Comment (ssoloff) [inline src/test/java/games/strategy/engine/framework/map/download/DownloadUtilsTests.java]: A nested fixture for the `downloadToFile()` method.

File: src/test/java/games/strategy/engine/framework/map/download/DownloadUtilsTests.java [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+      catchException(() -> downloadToFile());
+
+      assertThat(caughtException(), allOf(
+          is(instanceOf(IOException.class)),
+          hasMessageThat(containsString("entity is missing"))));
+    }
+  }
+
+  @RunWith(MockitoJUnitRunner.class)
+  public static final class GetDownloadLengthFromCache {
```
Comment (ssoloff) [inline src/test/java/games/strategy/engine/framework/map/download/DownloadUtilsTests.java]: A nested fixture for the `getDownloadLengthFromCache()` method.

File: src/test/java/games/strategy/engine/framework/map/download/DownloadUtilsTests.java [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+
+      final Optional<Long> downloadLength = getDownloadLengthFromCache();
+
+      assertThat(downloadLength, is(Optional.empty()));
+      assertThat(DownloadUtils.downloadLengthsByUri, is(anEmptyMap()));
+    }
+  }
+
+  @RunWith(MockitoJUnitRunner.class)
+  public static final class GetDownloadLengthFromHost {
```
Comment (ssoloff) [inline src/test/java/games/strategy/engine/framework/map/download/DownloadUtilsTests.java]: A nested fixture for the `getDownloadLengthFromHost()` method.

**Comment by ron-murhammer** [comment]
I'm fine with this. I'll let @DanVanAtta take a look as well.

**Comment by ssoloff** [comment]
I've been playing around with nested fixtures some more on an unpublished branch, and I'm now leaning towards including a `Test` suffix on the nested fixtures for various reasons, not the least of which is consistency with how we name other "leaf" fixtures.  Please feel free to yea or nay that idea before this PR gets merged.

**Comment by DanVanAtta** [comment]
@ssoloff - overall, is this your most preferred approach to solving the problem of grouping unit tests? I'm inclined to go this direction, I see benefits to this approach (well grouped, single self contained test files with reasonable/minimal configuration), but am curious if there are other preferred options beyond my thinking it's good to have a 1:1 ratio between unit tests and 'prod' classes.

**Comment by ssoloff** [comment]
@DanVanAtta I'm a big fan of BDD-style unit test frameworks and have been heavily influenced in the Javascript world by [Jasmine](https://jasmine.github.io/) (and subsequently Jest).  Being able to nest fixtures like this seems very natural and allows for much better reporting, IMO.

Basically, I've observed that many fixtures have the following form:

```
ClassOrModule
    MethodOrFeature
        WhenConditionOne
            ShouldDoA
            ShouldDoB
        WhenConditionTwo
            ShouldDoC
            ShouldDoD
```

So being able to structure it that way seems preferable to structuring it using a flat style:

```
ClassOrModule
    MethodOrFeature_ShouldDoA_WhenConditionOne
    MethodOrFeature_ShouldDoB_WhenConditionOne
    MethodOrFeature_ShouldDoC_WhenConditionTwo
    MethodOrFeature_ShouldDoD_WhenConditionTwo
```

In the Java BDD test framework world, I've used Spock pretty heavily on a project about a year ago, but was disappointed that it didn't support nested tests.  So I really have no other Java-specific examples to draw knowledge from regarding nesting fixtures.  (I never got on the TestNG train, so I don't know if they have their own solution.)

Other than that, the "classic" convention I've seen proposed by Kent Beck, Roy Osherove, and others is what I had done prior to this PR.  Namely, create multiple fixtures for each method/feature where it makes sense (i.e. when the state of the original fixture starts to carry orthogonal responsibilities).  However, the fixtures are **not** usually placed in separate files the way I did it.  Specifically, the examples I've seen recommend the second and subsequent fixtures be made (in the case of JUnit and Spock) top-level package-private classes in the same file.  Unfortunately, that's currently a Google Style violation, and I've since avoided it.

**Pull request merged**